### PR TITLE
[AUTO] Update AUTO logic to support state-full model

### PR DESCRIPTION
**Pull request by yangwang201911**
### Details:
 - Update logic to filter out the devices supported state-full model
 - Add test cases

### Tickets:
 - 84703

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@yangwang201911 @songbell could you please clarify whether the config for AUTO plugin is correct here https://github.com/openvinotoolkit/openvino/blob/master/.github/components.yml#L71 ?

It's smart CI config, where `revalidate` means that changes in AUTO component needs to run tests for `revalidate` components. I don't know whether changes in AUTO affects CPU plugin tests ? If no, we can remove GPU, CPU and some others from `revalidate` for AUTO

And, please, check the same for AUTO_BATCH

File: src/plugins/auto/src/plugin.cpp [hunk 1]
```diff
@@ -911,39 +911,69 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+                return false;
+        }
+        return true;
+    };
+
+    for (auto& item : meta_devices) {
+        if (is_supported_stateful(item.device_name, item.config))
+            filter_device.push_back(item);
+    }
+    if (get_device_name() == "MULTI") {
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:967]: how about auto ctput

File: src/plugins/auto/src/plugin.cpp [hunk 2]
```diff
@@ -911,39 +911,69 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+        if (filter_device.empty() || filter_device.size() > 1)
+            OPENVINO_THROW("AUTO cumulative model didn't support stateful model.");
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:969]:
```suggestion
            OPENVINO_THROW("AUTO cumulative mode doesn't support stateful model.");
```

File: src/plugins/auto/src/plugin.cpp [hunk 3]
```diff
@@ -911,39 +911,69 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
         filter_device.push_back(*cpuiter);
         return filter_device;
     }
-
     // If CPU is not in candidate list, continue to run selection logic regardless of whether the input model is a
     // dynamic model or not
-    return meta_devices;
+
+    for (auto& op : model->get_ops()) {
+        if (std::dynamic_pointer_cast<ngraph::op::AssignBase>(op) ||
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:938]: use new name space instead of ngraph

Comment (songbell) [reply src/plugins/auto/src/plugin.cpp:938]:
```suggestion
        if (is_type(ov::op::util::ReadValueBase) ||
```

File: src/plugins/auto/tests/unit/stateful_model_test.cpp [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    if (expectedTimes < 0) {
+        ASSERT_THROW(plugin->compile_model(model, config), ov::Exception);
+    } else {
+        ASSERT_NO_THROW(plugin->compile_model(model, config));
+    }
+}
+
+const std::vector<StatefulModelConfigParams> testConfigs = {
+    // test cases for dynamic model
+    StatefulModelConfigParams{"CPU",
```
Comment (wangleis) [inline src/plugins/auto/tests/unit/stateful_model_test.cpp:202]: Please add description for each item in first test data set.

File: src/plugins/auto/src/plugin.cpp (hunk 3 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp]: Updated.

File: src/plugins/auto/src/plugin.cpp (hunk 2 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp]: Updated.

File: src/plugins/auto/src/plugin.cpp (hunk 1 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp]: Updated.

File: src/plugins/auto/tests/unit/stateful_model_test.cpp (hunk 4 above)
Comment (yangwang201911) [reply src/plugins/auto/tests/unit/stateful_model_test.cpp]: Updated

File: src/plugins/auto/src/plugin.cpp [hunk 5]
```diff
@@ -911,39 +911,71 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+            stateful_node_names.push_back(op->get_friendly_name());
+        }
+    }
+    if (stateful_node_names.empty()) {
+        // not stateful model
+        return meta_devices;
+    }
+
+    // disable CPU_HELP if model is stateful
+    load_config.set_property(ov::intel_auto::enable_startup_fallback(false));
```
Comment (peterchen-intel) [inline src/plugins/auto/src/plugin.cpp:950]: We should disable ov::intel_auto::enable_runtime_fallback as well. TODO: enable enable_runtime_fallback when stateful reset_state is ready.

File: src/plugins/auto/src/plugin.cpp [hunk 6]
```diff
@@ -911,39 +911,71 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
     }
-
     // If CPU is not in candidate list, continue to run selection logic regardless of whether the input model is a
     // dynamic model or not
-    return meta_devices;
+
+    for (auto& op : model->get_ops()) {
+        if (std::dynamic_pointer_cast<ov::op::util::AssignBase>(op) ||
+            std::dynamic_pointer_cast<ov::op::util::ReadValueBase>(op)) {
+            LOG_INFO_TAG("stateful mode, try deployed to CPU");
```
Comment (ilya-lavrenov) [inline src/plugins/auto/src/plugin.cpp:940]: BTW, GPU also supports stateful models

**Review by ilya-lavrenov — COMMENTED** [review]
Common question: when I create multiple AUTO inference requests for stateful models, do we have a guarantee that when I call `req.infer()` multiple times, then it comes to the same HW infer request?

File: src/plugins/auto/src/plugin.cpp (hunk 6 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp:940]: Updated.

File: src/plugins/auto/src/plugin.cpp (hunk 5 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp:950]: Updated.

**Comment by peterchen-intel** [comment]
> Common question: when I create multiple AUTO inference requests for stateful models, do we have a guarantee that when I call `req.infer()` multiple times, then it comes to the same HW infer request?

Yes. Currently AUTO supports stateful model in single device scenario only.

**Comment by wangleis** [comment]
> Common question: when I create multiple AUTO inference requests for stateful models, do we have a guarantee that when I call `req.infer()` multiple times, then it comes to the same HW infer request?

@ilya-lavrenov Just to clarify the design of current implementation, AUTO will not support state-full model in CUMULATIVE THROUGHPUT mode when available candidate device is not than one device. In other mode, AUTO will disable startup fallback and runtime fallback when load state-full model to single device.

File: src/plugins/auto/src/plugin.cpp [hunk 7]
```diff
@@ -910,39 +910,77 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+    // disable CPU_HELP if model is stateful
+    if (load_config.get_property(ov::intel_auto::enable_startup_fallback)) {
+        LOG_WARNING_TAG("Setting property ov::intel_auto::enable_startup_fallback to false for stateful model.");
+        load_config.set_property(ov::intel_auto::enable_startup_fallback(false));
+    }
+    if (load_config.get_property(ov::intel_auto::enable_runtime_fallback)) {
+        LOG_WARNING_TAG("Setting property ov::intel_auto::enable_running_fallback to false for stateful model.");
+        load_config.set_property(ov::intel_auto::enable_runtime_fallback(false));
+    }
+    if (meta_devices.size() == 1)
```
Comment (wangleis) [inline src/plugins/auto/src/plugin.cpp:956]: Please check device size first

Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp]: updated.

File: src/plugins/auto/src/plugin.cpp [hunk 8]
```diff
@@ -910,39 +910,84 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+        }
+        if (load_config.get_property(ov::intel_auto::enable_runtime_fallback)) {
+            LOG_WARNING_TAG("Setting property ov::intel_auto::enable_running_fallback to false for stateful model.");
+            load_config.set_property(ov::intel_auto::enable_runtime_fallback(false));
         }
-        return false;
     };
 
+    if (meta_devices.size() == 1) {
+        disable_startup_runtime_fallback();
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:931]: if only one device, we don't need to disable this？

File: src/plugins/auto/src/plugin.cpp [hunk 9]
```diff
@@ -910,39 +910,84 @@ std::vector<DeviceInformation> Plugin::filter_device(const std::vector<DeviceInf
...
+            std::dynamic_pointer_cast<ov::op::util::ReadValueBase>(op)) {
+            stateful_node_names.push_back(op->get_friendly_name());
+        }
+    }
+    if (stateful_node_names.empty()) {
+        // not stateful model
+        return meta_devices;
+    }
+
+    // disable CPU_HELP if model is stateful
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:963]:
```suggestion
    // disable CPU_HELP and runtime fallback if model is stateful
```

**Review by songbell — APPROVED** [review]

File: src/plugins/auto/src/plugin.cpp (hunk 8 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp:931]: Yes. Found the similar logic for one device.

File: src/plugins/auto/src/plugin.cpp (hunk 9 above)
Comment (yangwang201911) [reply src/plugins/auto/src/plugin.cpp]: Updated.

**Review by wangleis — APPROVED** [review]

**Pull request merged**